### PR TITLE
modify the logging module

### PR DESCRIPTION
**Pull request by zzzzwj**
replace the manual logging module with python package "logging.Logger"

### CODE REVIEW

**Comment by xuyxu** [comment]
Hi @zzzzwj, thanks for the PR!

Could you provide a snippet of the logging results ?

**Comment by zzzzwj** [comment]
Of course, yes. Here is an example output logging by running examples/classification_cifar10_cnn.py.
image
Anyway you can define your own logger by configuring logging.Logger like that in utils.get_default_logger() if you don't like the logging format.

**Comment by xuyxu** [comment]
Looks great! I will take a look at your codes latter :-)

File: torchensemble/_base.py [hunk 1]
```diff
@@ -44,10 +44,10 @@ class BaseModule(abc.ABC, nn.Module):
     def __init__(self,
                  estimator,
                  n_estimators,
+                 logger,
```
Comment (xuyxu) [inline torchensemble/_base.py:47]: Could we change the location of ``logger`` to the place right below ``estimator_args`` ? I think it's better because ``logger``, ``cuda``, and ``n_jobs`` are all utility-related.

File: torchensemble/_base.py [hunk 2]
```diff
@@ -62,7 +62,7 @@ def __init__(self,
         self.estimator_args = estimator_args
         self.device = torch.device("cuda" if cuda else "cpu")
         self.n_jobs = n_jobs
-        self.verbose = verbose
+        self.logger = logger
```
Comment (xuyxu) [inline torchensemble/_base.py:65]: In addition, registering logger as an attribute should be above ``self.device``

File: torchensemble/_constants.py [hunk 3]
```diff
@@ -5,6 +5,8 @@
         The class of base estimator inherited from :mod:`torch.nn.Module`.
     n_estimators : int
         The number of base estimators in the ensemble.
+    logger : logging.Logger
+        The logger that records information generated by program.
```
Comment (xuyxu) [inline torchensemble/_constants.py:9]: Again, concerns on the place of ``logger``

File: torchensemble/bagging.py [hunk 4]
```diff
@@ -55,20 +55,23 @@ def _parallel_fit_per_epoch(train_loader,
...
             if is_classification:
                 pred = sampling_output.data.max(1)[1]
                 correct = pred.eq(sampling_target.view(-1).data).sum()
 
-                msg = ("{} Estimator: {:03d} | Epoch: {:03d} | Batch: {:03d}"
+                msg = ("Estimator: {:03d} | Epoch: {:03d} | Batch: {:03d}"
                        " | Loss: {:.5f} | Correct: {:d}/{:d}")
-                print(msg.format(utils.ctime(), idx, epoch, batch_idx, loss,
-                                 correct, sampling_data.size()[0]))
+                # print(msg.format(utils.ctime(), idx, epoch, batch_idx, loss,
```
Comment (xuyxu) [inline torchensemble/bagging.py:66]: We can remove all lines masked by ``#``

**Review by xuyxu — COMMENTED** [review]
* The location of the input argument ``logger``
* Remove lines masked by ``#``

**Comment by xuyxu** [comment]
Hi @zzzzwj, I have left several comments on your submitted code. 

Let me know if solving these comments would be better, and always open for discussion.

File: torchensemble/_base.py (hunk 1 above)
Comment (zzzzwj) [reply torchensemble/_base.py:47]: It's okay but I can't find out an elegant way to set the default value of `logger`. An default global logger may be created in `utils.py` to be assigned. I'll try to fix it in few days.

Comment (xuyxu) [reply torchensemble/_base.py:47]: I think a default logger is preferred, thanks!

**Comment by xuyxu** [comment]
Thanks, the logger looks much better now :-).

Here is a problem: It looks like that the logging results were printed twice, could you fix it ?

screenshot
